File: workers/download-tracker/src/pair.js
/**
 * AZNet sibling pair-status client.
 * Does not embed the AZNet protocol. Viewer + pair-status only.
 * FragGate unlocks; StaticClock times. Both products required.
 */

export const AZNET = "https://github.com/AzielEliab/aznet";
export const AZNET_WORKER = "https://aznet-download-tracker.vibelock.workers.dev";
export const AZNET_GARDEN = "https://aznet-download-tracker.vibelock.workers.dev/garden";
export const STATICCLOCK = "https://github.com/AzielEliab/staticclock";
export const FRAGGATE = "https://github.com/AzielEliab/fraggate";
export const FRAGGATE_CALL = "https://aziel-runtime.vibelock.workers.dev/v1/fraggate/call";
export const RUNTIME = "https://aziel-runtime.vibelock.workers.dev";

const UA = "Mozilla/5.0 AZBrowser/0.1.0 (pair-status; sibling-client)";

export function links() {
  return {
    aznet_github: AZNET,
    aznet_worker: AZNET_WORKER,
    aznet_garden: AZNET_GARDEN,
    fraggate: FRAGGATE,
    fraggate_call: FRAGGATE_CALL,
    runtime: RUNTIME,
    staticclock: STATICCLOCK,
  };
}

export function pairBase() {
  return {
    action: "pair_status",
    peer: "azbrowser",
    sibling: "aznet",
    required: true,
    viewer: true,
    protocol_embedded: false,
    links: links(),
    note: "AZBrowser views the AZNet side-net. Pairing is required to run. Protocol lives in AZNet. FragGate unlocks; StaticClock times.",
  };
}

export function stubPair(paired) {
  const base = pairBase();
  if (paired) {
    return {
      ...base,
      ok: true,
      paired: true,
      code: "PAIRED",
      aznet: { reachable: true, paired: true, source: "stub" },
      fraggate_unlock: true,
      staticclock: { source: "stub", note: "Tests do not call StaticClock." },
    };
  }
  return {
    ...base,
    ok: true,
    paired: false,
    code: "PAIR_REQUIRED",
    aznet: { reachable: false, paired: false, source: "stub" },
    fraggate_unlock: false,
    staticclock: {},
  };
}

async function getJson(url) {
  try {
    const res = await fetch(url, { headers: { "User-Agent": UA, Accept: "application/json" } });
    const text = await res.text();
    try {
      return { ok: res.ok, body: JSON.parse(text) };
    } catch {
      return { ok: res.ok, body: { ok: res.ok, text: text.slice(0, 200) } };
    }
  } catch (exc) {
    return { ok: false, body: { error: String(exc).slice(0, 240) } };
  }
}

export async function livePairStatus() {
  const base = pairBase();
  const health = await getJson(AZNET_WORKER.replace(/\/$/, "") + "/v1/health");
  const pair = await getJson(AZNET_WORKER.replace(/\/$/, "") + "/v1/pair?peer=azbrowser");
  const aznetReachable = Boolean(health.ok || pair.ok);
  const blob = pair.ok ? pair.body : health.body;
  let aznetPaired = false;
  if (blob && typeof blob === "object" && (blob.paired === true || (blob.ok === true && blob.peer === "azbrowser"))) {
    aznetPaired = true;
  }

  let fgOk = false;
  let sc = {};
  try {
    const res = await fetch(FRAGGATE_CALL, {
      method: "POST",
      headers: { "User-Agent": UA, "Content-Type": "application/json", Accept: "application/json" },
      body: JSON.stringify({ slug: "staticclock", op: "advise", payload: { geo: "UTC" } }),
    });
    const fg = await res.json();
    fgOk = Boolean(res.ok);
    const inner = (fg && fg.result && fg.result.result) || (fg && fg.result) || {};
    if (inner && typeof inner === "object") {
      sc = {
        optimal_time: inner.optimal_time,
        optimal_date: inner.optimal_date,
        motto: inner.motto,
        source: "fraggate/staticclock",
      };
    }
    fgOk = Boolean(fgOk && fg && (fg.ok === true || (fg.result && fg.result.ok === true) || sc.optimal_time || sc.motto));
  } catch (exc) {
    sc = { error: String(exc).slice(0, 240) };
    fgOk = false;
  }

  const paired = Boolean(aznetPaired && fgOk);
  return {
    ...base,
    ok: true,
    paired,
    code: paired ? "PAIRED" : "PAIR_REQUIRED",
    aznet: { reachable: aznetReachable, paired: aznetPaired, detail: blob, source: "worker" },
    fraggate_unlock: fgOk,
    staticclock: sc,
  };
}

export function sidenetView() {
  return {
    ok: true,
    action: "sidenet_view",
    viewer: true,
    protocol_embedded: false,
    garden: AZNET_GARDEN,
    worker: AZNET_WORKER,
    github: AZNET,
    links: links(),
    note: "AZNet branding here is the side-net viewer. The protocol and whitepaper live in https://github.com/AzielEliab/aznet.",
  };
}

export async function resolvePair(session, payload) {
  if (payload && typeof payload.paired === "boolean") session.pairStub = payload.paired;
  if (session.pairStub === true) return stubPair(true);
  if (session.pairStub === false) return stubPair(false);
  if (session.pairCache) return session.pairCache;
  const status = await livePairStatus();
  session.pairCache = status;
  return status;
}

File: azbrowser/pair.py
"""AZNet sibling pair-status client.

Does not embed the AZNet protocol. Talks to the documented Worker/garden
and records FragGate unlock + StaticClock time. Both products required.
"""

from __future__ import annotations

import json
from typing import Any
from urllib.request import Request, urlopen

from .meta import (
    AZNET,
    AZNET_GARDEN,
    AZNET_WORKER,
    FRAGGATE,
    FRAGGATE_CALL,
    RUNTIME,
    STATICCLOCK,
)

UA = "Mozilla/5.0 AZBrowser/0.1.0 (pair-status; sibling-client)"


def _get_json(url: str, timeout: float = 4.0) -> tuple[bool, Any]:
    try:
        req = Request(url, headers={"User-Agent": UA, "Accept": "application/json"})
        with urlopen(req, timeout=timeout) as resp:  # noqa: S310 — documented sibling/public URL
            raw = resp.read(64_000)
            try:
                return True, json.loads(raw.decode("utf-8", errors="replace"))
            except json.JSONDecodeError:
                return True, {"ok": resp.status == 200, "text": raw[:200].decode("utf-8", errors="replace")}
    except Exception as exc:
        return False, {"error": str(exc)[:240]}


def _post_json(url: str, body: dict[str, Any], timeout: float = 6.0) -> tuple[bool, Any]:
    try:
        req = Request(
            url,
            data=json.dumps(body).encode("utf-8"),
            headers={"User-Agent": UA, "Content-Type": "application/json", "Accept": "application/json"},
            method="POST",
        )
        with urlopen(req, timeout=timeout) as resp:  # noqa: S310
            return True, json.loads(resp.read(64_000).decode("utf-8", errors="replace"))
    except Exception as exc:
        return False, {"error": str(exc)[:240]}


def links() -> dict[str, str]:
    return {
        "aznet_github": AZNET,
        "aznet_worker": AZNET_WORKER,
        "aznet_garden": AZNET_GARDEN,
        "fraggate": FRAGGATE,
        "fraggate_call": FRAGGATE_CALL,
        "runtime": RUNTIME,
        "staticclock": STATICCLOCK,
    }


def pair_status(*, fetch: bool = True, stub_paired: bool | None = None) -> dict[str, Any]:
    """Return pair-status. stub_paired bypasses the network (tests / doctor)."""
    base = {
        "action": "pair_status",
        "peer": "azbrowser",
        "sibling": "aznet",
        "required": True,
        "viewer": True,
        "protocol_embedded": False,
        "links": links(),
        "note": "AZBrowser views the AZNet side-net. Pairing is required to run. Protocol lives in AZNet. FragGate unlocks; StaticClock times.",
    }
    if stub_paired is True:
        return {
            **base,
            "ok": True,
            "paired": True,
            "code": "PAIRED",
            "aznet": {"reachable": True, "paired": True, "source": "stub"},
            "fraggate_unlock": True,
            "staticclock": {"source": "stub", "note": "Tests do not call StaticClock."},
        }
    if stub_paired is False:
        return {
            **base,
            "ok": True,
            "paired": False,
            "code": "PAIR_REQUIRED",
            "aznet": {"reachable": False, "paired": False, "source": "stub"},
            "fraggate_unlock": False,
            "staticclock": {},
        }

    aznet_reachable = False
    aznet_paired = False
    aznet_detail: Any = None
    if fetch:
        ok_h, health = _get_json(AZNET_WORKER.rstrip("/") + "/v1/health")
        ok_p, pair = _get_json(AZNET_WORKER.rstrip("/") + "/v1/pair?peer=azbrowser")
        aznet_reachable = bool(ok_h or ok_p)
        blob = pair if ok_p else health
        aznet_detail = blob
        if isinstance(blob, dict) and (blob.get("paired") is True or (blob.get("ok") is True and blob.get("peer") == "azbrowser")):
            aznet_paired = True

    fg_ok, fg = _post_json(FRAGGATE_CALL, {"slug": "staticclock", "op": "advise", "payload": {"geo": "UTC"}}) if fetch else (False, {})
    sc = {}
    if fg_ok and isinstance(fg, dict):
        inner = (fg.get("result") or {}).get("result") or fg.get("result") or {}
        if isinstance(inner, dict):
            sc = {
                "optimal_time": inner.get("optimal_time"),
                "optimal_date": inner.get("optimal_date"),
                "motto": inner.get("motto"),
                "source": "fraggate/staticclock",
            }
    fraggate_unlock = bool(fg_ok and isinstance(fg, dict) and (fg.get("ok") is True or (fg.get("result") or {}).get("ok") is True or sc))

    paired = bool(aznet_paired and fraggate_unlock)
    return {
        **base,
        "ok": True,
        "paired": paired,
        "code": "PAIRED" if paired else "PAIR_REQUIRED",
        "aznet": {"reachable": aznet_reachable, "paired": aznet_paired, "detail": aznet_detail, "source": "worker"},
        "fraggate_unlock": fraggate_unlock,
        "staticclock": sc,
    }


def sidenet_view() -> dict[str, Any]:
    """Deep-link viewer envelope. Not an AZNet protocol implementation."""
    return {
        "ok": True,
        "action": "sidenet_view",
        "viewer": True,
        "protocol_embedded": False,
        "garden": AZNET_GARDEN,
        "worker": AZNET_WORKER,
        "github": AZNET,
        "links": links(),
        "note": "AZNet branding here is the side-net viewer. The protocol and whitepaper live in https://github.com/AzielEliab/aznet.",
    }
